cpp/src/arrow/io/io-compressed-test.cc

```diff
@@ -0,0 +1,243 @@
+  ASSERT_OK(RunCompressedInputStream(codec, compressed, &decompressed));
+
+  ASSERT_EQ(decompressed.size(), data.size());
+  ASSERT_EQ(decompressed, data);
+}
+
+void CheckCompressedOutputStream(Codec* codec, const std::vector<uint8_t>& data,
+                                 bool do_flush) {
+  // Create compressed output stream
+  std::shared_ptr<BufferOutputStream> buffer_writer;
+  ASSERT_OK(BufferOutputStream::Create(1024, default_memory_pool(), &buffer_writer));
+  std::shared_ptr<CompressedOutputStream> stream;
+  ASSERT_OK(CompressedOutputStream::Make(codec, buffer_writer, &stream));
+
+  const uint8_t* input = data.data();
+  int64_t input_len = data.size();
+  const int64_t chunk_size = 1111;
+  while (input_len > 0) {
+    int64_t nbytes = std::min(chunk_size, input_len);
+    ASSERT_OK(stream->Write(input, nbytes));
+    input += nbytes;
+    input_len -= nbytes;
+    if (do_flush) {
+      ASSERT_OK(stream->Flush());
+    }
+  }
+  ASSERT_OK(stream->Close());
+
+  // Get compressed data and decompress it
+  std::shared_ptr<Buffer> compressed;
+  ASSERT_OK(buffer_writer->Finish(&compressed));
+  std::vector<uint8_t> decompressed(data.size());
+  ASSERT_OK(codec->Decompress(compressed->size(), compressed->data(), decompressed.size(),
+                              decompressed.data()));
+  ASSERT_EQ(decompressed, data);
+}
+
+class CompressedInputStreamTest : public ::testing::TestWithParam<Compression::type> {
+ protected:
+  Compression::type GetCompression() { return GetParam(); }
+
+  std::unique_ptr<Codec> MakeCodec() {
+    std::unique_ptr<Codec> codec;
+    ABORT_NOT_OK(Codec::Create(GetCompression(), &codec));
+    return codec;
+  }
+};
+
+TEST_P(CompressedInputStreamTest, CompressibleData) {
+  auto codec = MakeCodec();
+  auto data = MakeCompressibleData(COMPRESSIBLE_DATA_SIZE);
+
+  CheckCompressedInputStream(codec.get(), data);
+}
+
+TEST_P(CompressedInputStreamTest, RandomData) {
+  auto codec = MakeCodec();
+  auto data = MakeRandomData(RANDOM_DATA_SIZE);
+
+  CheckCompressedInputStream(codec.get(), data);
+}
+
+TEST_P(CompressedInputStreamTest, TruncatedData) {
+  auto codec = MakeCodec();
+  auto data = MakeRandomData(10000);
+  auto compressed = CompressDataOneShot(codec.get(), data);
+  auto truncated = SliceBuffer(compressed, 0, compressed->size() - 3);
+
+  std::vector<uint8_t> decompressed;
+  ASSERT_RAISES(IOError, RunCompressedInputStream(codec.get(), truncated, &decompressed));
+}
+
+TEST_P(CompressedInputStreamTest, InvalidData) {
+  auto codec = MakeCodec();
+  auto compressed_data = MakeRandomData(10000);
+
+  auto buffer_reader = std::make_shared<BufferReader>(Buffer::Wrap(compressed_data));
+  std::shared_ptr<CompressedInputStream> stream;
+  ASSERT_OK(CompressedInputStream::Make(codec.get(), buffer_reader, &stream));
+  std::shared_ptr<Buffer> out_buf;
+  ASSERT_RAISES(IOError, stream->Read(1024, &out_buf));
+}
+
+// NOTE: Snappy doesn't support streaming decompression
```

Comment:
We could (should?) define a framed format of our own devising

Reply:
But is it useful? The main point here is to interact with existing files.